File: 216.组合总和-iii.py
#
# @lc app=leetcode.cn id=216 lang=python3
#
# [216] 组合总和 III
#
# https://leetcode-cn.com/problems/combination-sum-iii/description/
#
# algorithms
# Medium (70.55%)
# Likes:    112
# Dislikes: 0
# Total Accepted:    20.5K
# Total Submissions: 28.9K
# Testcase Example:  '3\n7'
#
# 找出所有相加之和为 n 的 k 个数的组合。组合中只允许含有 1 - 9 的正整数，并且每种组合中不存在重复的数字。
# 
# 说明：
# 
# 
# 所有数字都是正整数。
# 解集不能包含重复的组合。 
# 
# 
# 示例 1:
# 
# 输入: k = 3, n = 7
# 输出: [[1,2,4]]
# 
# 
# 示例 2:
# 
# 输入: k = 3, n = 9
# 输出: [[1,2,6], [1,3,5], [2,3,4]]
# 
# 
#

# @lc code=start
class Solution:
    def backtrack(self, start, path:list, k, n, ans:list):
        if k<=0 or n<=0:
            if k==0 and n==0:
                ans.append(path.copy())
            return
        for i in range(start,10):
            if n-i>=0:
                path.append(i)
                self.backtrack(i+1, path, k-1, n-i, ans)
                path.pop()
        

    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ans = []
        self.backtrack(1, [], k, n, ans)
        return ans
# @lc code=end

File: newlife/combination-sum/216.组合总和-iii.cpp
/*
 * @lc app=leetcode.cn id=216 lang=cpp
 *
 * [216] 组合总和 III
 *
 * https://leetcode.cn/problems/combination-sum-iii/description/
 *
 * algorithms
 * Medium (72.71%)
 * Likes:    672
 * Dislikes: 0
 * Total Accepted:    262.3K
 * Total Submissions: 366.2K
 * Testcase Example:  '3\n7'
 *
 * 找出所有相加之和为 n 的 k 个数的组合，且满足下列条件：
 * 
 * 
 * 只使用数字1到9
 * 每个数字 最多使用一次 
 * 
 * 
 * 返回 所有可能的有效组合的列表 。该列表不能包含相同的组合两次，组合可以以任何顺序返回。
 * 
 * 
 * 
 * 示例 1:
 * 
 * 
 * 输入: k = 3, n = 7
 * 输出: [[1,2,4]]
 * 解释:
 * 1 + 2 + 4 = 7
 * 没有其他符合的组合了。
 * 
 * 示例 2:
 * 
 * 
 * 输入: k = 3, n = 9
 * 输出: [[1,2,6], [1,3,5], [2,3,4]]
 * 解释:
 * 1 + 2 + 6 = 9
 * 1 + 3 + 5 = 9
 * 2 + 3 + 4 = 9
 * 没有其他符合的组合了。
 * 
 * 示例 3:
 * 
 * 
 * 输入: k = 4, n = 1
 * 输出: []
 * 解释: 不存在有效的组合。
 * 在[1,9]范围内使用4个不同的数字，我们可以得到的最小和是1+2+3+4 = 10，因为10 > 1，没有有效的组合。
 * 
 * 
 * 
 * 
 * 提示:
 * 
 * 
 * 2 <= k <= 9
 * 1 <= n <= 60
 * 
 * 
 */
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    void backtrack(vector<vector<int>>& ans,vector<int>&& path, int start, int target, int k){
        if(target==0 && path.size()==k){
            ans.push_back(path);
            return;
        }
        if(target<0||path.size()>k) return;
        for(int i=start;i<10;i++) {
            path.push_back(i);
            backtrack(ans, std::move(path), i+1, target-i, k);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        backtrack(ans, {}, 1, n, k);
        return ans;
    }
};
// @lc code=end

